### PR TITLE
Feat PyTorch calibrator and two-phase training pipeline

### PR DESCRIPTION
**Pull request by JemmaLDaniel**
## Summary

Replaces the scikit-learn `MLPClassifier` calibrator with a PyTorch `CalibratorNetwork`, saved as `model.safetensors` + `config.json`. Adds a scalable two-phase train path (compute features to Parquet, then `fit_from_features`) and revises compute-features / predict / diagnose-calibration CLI orchestration to match the `revisions` branch pipeline.

### Calibrator

- PyTorch MLP with configurable `hidden_dims`, dropout, Adam training, GPU when available
- Early stopping: `n_iter_no_change` + `tol`; optional validation subsampling (`val_early_stopping_max_psms`)
- `TrainingHistory` (epoch metrics, JSON export, plot)
- `fit()` / `fit_from_features()`; `predict()` expects pre-computed features (no internal `compute_features`)
- Save/load via safetensors; feature configs in JSON (`get_config()`); Koina CE/frag runtime keys stripped on save
- Predict-time merge of `koina.input_constants` / `input_columns` onto loaded models

### Data & training

- `FeatureDataset` + `from_parquet()` (file or directory, optional `feature_columns`)
- `winnow train`: single-phase (features on full set → split → fit) or two-phase (`features_path` / `val_features_path`)
- `resolve_data_path()` for local paths and HuggingFace Hub models/datasets

### CLI & config

- Batched `compute-features` / `predict` (per-experiment directories, shared filtering)
- `configs/koina.yaml` extracted; slim `predict.yaml` / `diagnose_calibration.yaml` (no full train calibrator tree included where unnecessary)
- `irt_calibration` predict overrides; hybrid iRT (skip/warn on bad pools instead of raising to keep training runs going, keep #188 loaded-regressor pinning)
- Sample `Makefile` targets use `n_iter_no_change`

## Test plan

- [x] `make tests` / pre-commit
- [x] `make train-sample` and `make predict-sample`
- [x] Calibrator: save/load roundtrip, Koina override merge, early stopping, `fit_from_features`, empty-dataset guards
- [x] FeatureDataset: Parquet dir + `feature_columns`
- [x] iRT: per-experiment fit, loaded-regressor skip, low-diversity warning
- [x] `test_koina_intensity_config`, `test_paths` (HF path resolution)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -22,6 +22,9 @@ dependencies = [
     "matchms>=0.31.0",
     "scikit-learn>=1.3.0",
     "matplotlib>=3.7.0",
+    "torch>=2.10.0",
```
Comment (BioGeek) [inline pyproject.toml]: I would recommend keeping the same `torch`[ version ranges as InstaNovo](https://github.com/instadeepai/InstaNovo/blob/main/pyproject.toml#L102-L112).

Now Winnow is at `"torch>=2.10.0"` and InstaNovo at `<2.5/<2.!/<2.9` (depending on the paltform and CUDA version), which makes it impossible to install Winnow and InstaNovo in the same environment or Google Colab.

File: docs/configuration.md [hunk 2]
```diff
@@ -221,51 +240,52 @@ calibrator:
 
     beam_features:
       _target_: winnow.calibration.calibration_features.BeamFeatures
+```
+
+Koina model names, collision-energy / fragmentation inputs, and validity constraints
+live in [`configs/koina.yaml`](configs/koina.yaml) (composed via `defaults: - koina` in
```
Comment (BioGeek) [inline docs/configuration.md]: Running `make docs` retuns:

```
uv run mkdocs build --strict

 │  ⚠  Warning from the Material for MkDocs team
 │
 │  MkDocs 2.0, the underlying framework of Material for MkDocs,
 │  will introduce backward-incompatible changes, including:
 │
 │  × All plugins will stop working – the plugin system has been removed
 │  × All theme overrides will break – the theming system has been rewritten
 │  × No migration path exists – existing projects cannot be upgraded
 │  × Closed contribution model – community members can't report bugs
 │  × Currently unlicensed – unsuitable for production use
 │
 │  Our full analysis:
 │
 │  https://squidfunk.github.io/mkdocs-material/blog/2026/02/18/mkdocs-2.0/

INFO    -  Cleaning site directory
INFO    -  Building documentation to directory: /home/j-vangoey/code/winnow/docs_public
WARNING -  Doc file 'configuration.md' contains a link 'configs/koina.yaml', but the target is not found among documentation files.

Aborted with 1 warnings in strict mode!
make: *** [Makefile:122: docs] Error 1
```

MkDocs resolves relative links only within the documentation source tree, not arbitrary repo files. Since `configs/koina.yaml` is not present as a docs page/file under the configured `docs` directory, MkDocs reports a broken link. Because `--strict` is enabled, that warning aborts the build.

The Material/MkDocs 2.0 message is just an upstream warning banner, not the build-breaking error. But something we propably will have to look into.

Likely fix is to use a full GitHub URL.

File: winnow/calibration/features/retention_time.py [hunk 3]
```diff
@@ -295,18 +294,114 @@ def compute(self, dataset: CalibrationDataset) -> None:
...
+        per_exp_train: Dict[str, pd.DataFrame] = {}
+        for exp_name, group in experiments_to_fit.items():
+            try:
+                train_data = self._select_training_data(group, exp_name)
+            except ValueError as e:
+                outcome = (
+                    "dropped" if not self.learn_from_missing else "imputed as zero"
+                )
+                warnings.warn(
+                    f"Skipping experiment '{exp_name}':\n{e}\n"
```
Comment (BioGeek) [inline winnow/calibration/features/retention_time.py]: When the input has no `experiment_name` and the global iRT regressor is skipped (for example because there are fewer than `min_train_points ` or fewer than two unique peptides), `_collect_training_data` records `"__global__"` in `_skipped_experiments`, but this branch only applies skips when an `experiment_name` column exists.  `compute()` then keeps those rows as  valid and `_apply_regressors()` indexes `self.irt_predictors["__global__"]`, raising `KeyError` instead of imputing or dropping the spectra as the warning says.

To reproduce: [`test_compute_imputes_skipped_global_regressor`](https://github.com/instadeepai/winnow/commit/681679cb5a300edc07851717ad1c0d0cf1cbb1f2)

File: winnow/scripts/main.py [hunk 4]
```diff
@@ -223,22 +567,164 @@ def train_entry_point(
...
+    predictions_path,
+    data_loader,
+    calibrator,
+    labelled: bool,
+) -> list[pd.DataFrame]:
+    """Run feature computation over a directory (one file at a time) or a single spectrum file."""
+    if not spectrum_path.exists():
+        raise FileNotFoundError(f"Spectrum path does not exist: {spectrum_path}")
+    if spectrum_path.is_dir():
+        return _compute_features_directory(
```
Comment (BioGeek) [inline winnow/scripts/main.py:755]: With `data_loader=winnow`, the valid input is a saved Winnow dataset directory containing `metadata.csv` and optionally `predictions.pkl `. This branch intercepts every directory and sends it to `_discover_experiment_files`, which only accepts  `.parquet `,  `.ipc`, or  `.mgf` files, so `predict`/`compute-features` now fail before  `WinnowDatasetLoader.load()` can read a previously saved Winnow dataset directory.

To reproduce, see: [`test_compute_features_loads_saved_winnow_dataset_directory`](https://github.com/instadeepai/winnow/commit/8b8834deff50d2d33d7938a881eda3a735125084)

**Review by BioGeek — CHANGES_REQUESTED** [review]

File: winnow/utils/rich_console.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+from rich.console import Console
+from rich.logging import RichHandler
+
+# Shared stderr console for CLI error messages.
+STDERR_CONSOLE = Console(stderr=True)
+
+
+def notebook_safe_rich_handler(**kwargs: object) -> RichHandler:
+    """Return a Rich log handler that uses ANSI colours, not OSC-8 hyperlinks.
```
Comment (BioGeek) [inline winnow/utils/rich_console.py:13]: We might borrow this for InstaNovo as well!

File: winnow/scripts/main.py [hunk 6]
```diff
@@ -182,38 +314,250 @@ def train_entry_point(
...
+    """Save iRT regressors and training history if configured.
+
+    Args:
+        cfg: Resolved Hydra config.
+        calibrator: The fitted calibrator.
+        history: TrainingHistory returned by fit().
+    """
     irt_regressor_output_path = cfg.get("irt_regressor_output_path")
     if irt_regressor_output_path:
         from winnow.calibration.calibration_features import RetentionTimeFeature
```
Comment (BioGeek) [inline winnow/scripts/main.py]: A single calibrator/`RetentionTimeFeature` instance is reused per-file in directory mode, so  `_skipped_experiments ` accumulates. If file A skips an experiment and a later file B reuses that experiment_name, B's valid iRT features are silently marked  `is_missing_irt_error=True ` and imputed as zero (or dropped) → wrong calibrated scores, silently.

File: winnow/calibration/calibrator.py [hunk 7]
```diff
@@ -186,91 +528,451 @@ def add_feature(self, feature: CalibrationFeatures) -> None:
...
+            )
+
+            if epoch_callback is not None:
+                assert history.val_losses is not None
+                epoch_callback(epoch, history.val_losses[-1])
+
+            if should_stop:
+                break
+
+        history.epochs_trained = epoch + 1  # noqa: F821 — loop always runs
```
Comment (BioGeek) [inline winnow/calibration/calibrator.py]: `history.epochs_trained = epoch + 1 ` references the loop variable after the for loop. With `max_epochs=0` (a valid, unvalidated int) the loop never runs, epoch is unbound and the result is an unhandled `NameError  instead of a clear config error.

File: winnow/calibration/calibrator.py [hunk 8]
```diff
@@ -186,91 +528,451 @@ def add_feature(self, feature: CalibrationFeatures) -> None:
...
+
+        Returns:
+            Epoch-level training metrics.
+        """
+        return self._fit_from_features(
+            train_dataset, val_dataset, progress_bar, epoch_callback
+        )
+
+    @torch.no_grad()
+    def predict(self, dataset: CalibrationDataset) -> None:
```
Comment (BioGeek) [inline winnow/calibration/calibrator.py:693]: Training uses batched DataLoaders but prediction materializes the entire feature tensor + all activations at once. On millions of PSMs this OOMs VRAM (CUDA OOM) or exhausts host RAM.

File: winnow/calibration/calibrator.py [hunk 9]
```diff
@@ -115,64 +396,125 @@ def load(
             Path, str
         ] = "InstaDeepAI/winnow-general-model",
         cache_dir: Optional[Path] = None,
-    ) -> "ProbabilityCalibrator":
-        """Load a pretrained calibrator from a local path or Hugging Face repository. If the path is a local directory path, it will be used directly. If it is a Hugging Face repository identifier, it will be downloaded from Hugging Face.
+    ) -> ProbabilityCalibrator:
+        """Load a calibrator from a local directory or HuggingFace Hub repo.
+
+        Expects the directory to contain ``model.safetensors`` and
```
Comment (BioGeek) [inline winnow/calibration/calibrator.py:426]: `load()` no longer detects legacy pickle (`calibrator.pkl`) checkpoints. Pointing at a pre-PyTorch checkpoint dir now yields a bare `FileNotFoundError`. Raise a proper error that the old format is no longer supported.

File: winnow/scripts/main.py
Comment (BioGeek) [inline winnow/scripts/main.py:1362]: `diagnose_calibration` loads via a single `data_loader.load(**dataset_params)` while `predict` streams directories one-at-a-time via `_compute_features_batched_metadata`. The same directory dataset that works under predict either errors or loads everything into RAM under `diagnose-calibration`.

File: winnow/scripts/main.py [hunk 10]
```diff
@@ -182,38 +314,250 @@ def train_entry_point(
...
+        "and validation sets. For rigorous evaluation, provide separate "
+        "train/val Parquet files via features_path and val_features_path.",
+        validation_fraction * 100,
+    )
+
+    n = len(train_dataset)
+    n_val = max(1, int(n * validation_fraction))
+    n_train = n - n_val
+
+    generator = torch.Generator().manual_seed(cfg.get("calibrator", {}).get("seed", 42))
```
Comment (BioGeek) [inline winnow/scripts/main.py]: `_maybe_split_validation` uses `torch.Generator().manual_seed(seed)` while `_maybe_split_calibration_dataset` uses [`np.random.default_rng(seed)`](https://github.com/instadeepai/winnow/blob/8b8834deff50d2d33d7938a881eda3a735125084/winnow/scripts/main.py#L498). Same seed will result in different orderings, so single-phase vs two-phase training produce different train/val partitions and can't be reproduced across modes. Factor into one shared index-generating helper.

File: winnow/scripts/main.py [hunk 11]
```diff
@@ -182,38 +314,250 @@ def train_entry_point(
...
+        val_data: Optional validation CalibrationDataset with feature columns,
+            or ``None``.
+
+    Returns:
+        TrainingHistory from the calibrator.
+    """
+    import numpy as np
+    from winnow.datasets.feature_dataset import FeatureDataset
+
+    features, labels = calibrator._extract_feature_matrix(train_data, labelled=True)
```
Comment (BioGeek) [inline winnow/scripts/main.py]: `_fit_from_calibration_datasets` duplicates the feature-extraction + FeatureDataset wrapping from `fit()`, reaching into the private `_extract_feature_matrix`. If that return contract changes, `fit()` picks it up but the CLI copy silently doesn't.

**Review by BioGeek — CHANGES_REQUESTED** [review]

File: docs/configuration.md (hunk 2 above)
Comment (JemmaLDaniel) [reply docs/configuration.md]: Fixed with GitHub URL

File: winnow/calibration/features/retention_time.py (hunk 3 above)
Comment (JemmaLDaniel) [reply winnow/calibration/features/retention_time.py]: This is fixed. When the global RT -> iRT regressor is skipped and there is no `experiment_name` column, `_mark_missing_spectra` now marks all spectra as missing if `__global__` is in `_skipped_experiments`, so `compute()` imputes them (or drops them when `learn_from_missing=False`) instead of reaching `_apply_regressors()` and raising a `KeyError`. The test now passes.

File: winnow/utils/rich_console.py (hunk 5 above)
Comment (JemmaLDaniel) [reply winnow/utils/rich_console.py:13]: Thanks! Something I did for the PAASTA practical

File: winnow/scripts/main.py (hunk 4 above)
Comment (JemmaLDaniel) [reply winnow/scripts/main.py:755]: Fixed, `_compute_features_batched_metadata` now detects saved Winnow dataset directories (those containing `metadata.csv`) and passes them straight to `_compute_features_single_file`

File: winnow/scripts/main.py (hunk 6 above)
Comment (JemmaLDaniel) [reply winnow/scripts/main.py]: Should be fixed with a `_skipped_experiments` reset in `prepare()` now, and I've added a test to catch it in future: `test_skipped_experiments_reset_between_prepare_calls`

File: winnow/calibration/calibrator.py (hunk 7 above)
Comment (JemmaLDaniel) [reply winnow/calibration/calibrator.py]: Added a catch and a test `test_max_epochs_must_be_positive`

File: winnow/calibration/calibrator.py (hunk 9 above)
Comment (JemmaLDaniel) [reply winnow/calibration/calibrator.py:426]: Added an error to handle this and a test `test_load_legacy_pickle_checkpoint_raises`

File: winnow/scripts/main.py (hunk 10 above)
Comment (JemmaLDaniel) [reply winnow/scripts/main.py]: Refactored, the two modes should now agree for the same seed and dataset size

File: winnow/calibration/calibrator.py (hunk 8 above)
Comment (JemmaLDaniel) [reply winnow/calibration/calibrator.py:693]: Should be addressed in commit [2f91ad3](https://github.com/instadeepai/winnow/pull/190/commits/2f91ad3ce8cf80e990f2acd8c56fc0201ef9f5c1), where I batch the feature matrix during the forward pass to solve the activation OOM issue.

File: winnow/scripts/main.py
Comment (JemmaLDaniel) [reply winnow/scripts/main.py:1362]: Addressed in [78512b0](https://github.com/instadeepai/winnow/pull/190/commits/78512b0e07e3ab0cccfad139868ab31809946e36) so that diagnose uses the same helper as predict.

File: winnow/scripts/main.py (hunk 11 above)
Comment (JemmaLDaniel) [reply winnow/scripts/main.py]: Addressed in [c0169e4](https://github.com/instadeepai/winnow/pull/190/commits/c0169e4b3884af1c31b740ecf67b0cc7a62b9e25). I've made a public `to_feature_dataset()` method to formalise the `CalibrationDataset` to `FeatureDataset` step, that both `fit()` and the CLI path now use.

**Review by BioGeek — APPROVED** [review]
Thanks for addressing all review comments. Looks good now!
